comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -0,0 +1,184 @@
+                }
+                Ok(false)
+            },
+            Err(e) => match e {
+                StorageError::UniqueViolation(_) => match self.update_number_of_hits(body_hash_string).await {
+                    Ok(_) => Ok(true),
+                    Err(e) => Err(e),
+                },
+                _ => Err(e),
+            },
+        }
+    }
+
+    /// Trims the dedup cache to the configured limit by removing the oldest entries
+    pub async fn truncate(&self) -> Result<usize, StorageError> {
+        let capacity = self.capacity;
+        self.connection
+            .with_connection_async(move |conn| {
+                let mut num_removed = 0;
+                let msg_count = dedup_cache::table
+                    .select(dsl::count(dedup_cache::id))
+                    .first::<i64>(conn)? as usize;
+                // Hysteresis added to minimize database impact
+                if msg_count > capacity {
+                    let remove_count = msg_count - capacity;
+                    let message_ids: Vec<i32> = dedup_cache::table
+                        .select(dedup_cache::id)
+                        .order_by(dedup_cache::stored_at.asc())
+                        .limit(remove_count as i64)
+                        .get_results(conn)?;
+                    num_removed = diesel::delete(dedup_cache::table)
+                        .filter(dedup_cache::id.eq_any(message_ids))
+                        .execute(conn)?;
+                }
+                debug!(
+                    target: LOG_TARGET,
+                    "Message dedup cache: msg_count {}, capacity {}, num_removed {}", msg_count, capacity, num_removed,
+                );
+                Ok(num_removed)
+            })
+            .await
+    }
+
+    async fn insert_body_hash(&self, body_hash: String) -> Result<usize, StorageError> {
+        self.connection
+            .with_connection_async(move |conn| {
+                let insert_result = diesel::insert_into(dedup_cache::table)
+                    .values(UpdateDedupCacheSql {
+                        body_hash: Some(body_hash.clone()),
+                        number_of_hits: Some(1),
+                        last_hit_at: Some(DedupCacheDatabase::formatted_naive_date_time()?),
+                    })
+                    .execute(conn)
+                    .map_err(Into::into);
+                match insert_result {
+                    Ok(val) => Ok(val),
+                    Err(diesel::result::Error::DatabaseError(kind, e_info)) => match kind {
+                        DatabaseErrorKind::UniqueViolation => Err(StorageError::UniqueViolation(body_hash)),
+                        _ => Err(diesel::result::Error::DatabaseError(kind, e_info).into()),
+                    },
+                    Err(e) => Err(e.into()),
+                }
+            })
+            .await
+    }
+
+    async fn update_number_of_hits(&self, body_hash: String) -> Result<usize, StorageError> {
```

Comment:
nit: this method does not take in a parameter, so increment is more accurate than update

```suggestion
    async fn increment_number_of_hits(&self, body_hash: String) -> Result<usize, StorageError> {
```